Add a first cut at an image generation script

jbewing opened the pull request:
This PR adds a script that processes an arbitrary image for the PoV display and outputs the necessary C code such that the processed data can be shown on the MCU.

I apologize for writing this. The algorithm is generally as follows at the moment:
- Take the segment of pixels in the right center of a thumbnailed view of the original image
- Save those
- Rotate the image 3 degrees (relative to the OG image i.e. only one "rotation" operation is performed from the base image to prevent tearing from successive rotation transformations)
- loop until we've gone 360 degrees
- output these processed images so that they can be copy-pasta'd into the MCU source

##### Usage

```bash
poetry run python image_processor.py <path_to_image>
# Example:
poetry run python image_processor.py /Users/jbewing/Desktop/a_golden_retriever.jpeg
# Most likely worth it to redirect output to a file as they get _quite_ large
run python image_processor.py /Users/jbewing/Desktop/a_golden_retriever.jpeg > /Users/jbewing/Desktop/golden_retriever_pov_pixels.txt
```

cc @DarkAce65 @brodigan-e 
Related to: https://github.com/brodigan-e/capstone-POV/issues/9

jbewing committed:
Add a first cut at an image gen script

- This script will let us process arbitary images and add them to the program memory of the MCU

jbewing committed:
Update processor to rotate a copied-view of the OG image

jbewing committed:
Update image processor to initialize each list as differentize

python reuses the same reference if intialized in one place like that which led to each row of pixels having a length of like 4k instead of 120 i.e. (35 * 120)

DarkAce65 commented on server/image_processor.py:
```diff
@@ -0,0 +1,38 @@
+    print('.ledValues = {\n')
+    for row in rgb_data:
+        crgbSrings = list(map(lambda pixel: f"CRGB({pixel[0]},{pixel[1]},{pixel[2]})", row))
+        print(f"{{{','.join(crgbSrings)}}},")
```
Is this comma here at the end the bug you were talking about?

jbewing replied:
Yeah this is the one. Can't believe I naively missed this. I fixed by adding a `join` instead of blindly concatenating each row with a comma.

DarkAce65 commented on server/image_processor.py:
```diff
@@ -0,0 +1,38 @@
+
+
+if __name__ == '__main__':
+    thePixels = process_image_for_mcu(sys.argv[1])
```
nit:
```suggestion
    the_pixels = process_image_for_mcu(sys.argv[1])
```

DarkAce65 commented on server/image_processor.py:
```diff
@@ -0,0 +1,38 @@
+    for theta in range(angular_segments_count):
+        copied_image = image.copy()
+        rotated_image = copied_image.rotate((360/angular_segments_count) * theta)
+        pixels = list(rotated_image.getdata())[(35 * 17):(35 * 17) + 35]
```
Are these 35s supposed to be pulled from `size`?

jbewing replied:
Ohh yeah this is my haste and partially due to me not being able to coerce python into doing what I wanted with a float <=> integer number auto-casting. I'll extract these magic numbers into named constants.

jbewing replied:
Also, just so this is documented somewhere the overall rationale here is:

```
0-indexed 2D color matrix of a 35x35 image 
========================================
| (0, 0)  . . . (0, 17) . . .  (0, 34) |
|    .              .            .     |
|    .              .            .     |
|    .              .            .     |
| (17, 0) . . . (17, 17) . . . (17, 34)|
|    .              .             .    |
|    .              .             .    |
|    .              .             .    |
| (34, 0) . . . (34, 17) . . . (34, 34)|
========================================

- 17 is significant as `17 = floor(35/2)`
```

We're trying to extract the range of pixels in the middle from `(17, 0)` to `(17, 34)`. When the row-major 2D matrix is flattened to a list, these pixels should be in the range indexed by `(17, 0) => 35*17 + 0` to `(17, 34) => 35*17 + 34`. (One of the arms will display `(17,0) ---> (17, 17)` and the other will display `(17, 17) ---> (17, 34)`)

If we extract this range each time the image is rotated 3 degrees, then we should have a _nice_ contiguous view of the image through it's entire rotation.

jbewing committed:
Update server/image_processor.py

`thePixels`->`the_pixels` (more python-ey)

Co-authored-by: Taha Vasowalla <tvdood99@gmail.com>

jbewing committed:
PR feedback

- named constants
- fix bug with redundant comma printed after last rgb array literal
